add scorecard custom test proposal version 2 (#2687)
operator-framework/operator-sdk: doc/proposals/scorecard-custom-tests-2.md

Scorecard users will be able to construct custom tests and execute them using the scorecard.

#### Story 2

Scorecard will ship with an example custom test that end users could use as a reference when writing their custom tests.

#### Story 3

Scorecard internal tests will migrate to the custom test format and therefore be externalized from the scorecard binary itself.

#### Story 4 (optional)

Scorecard would support the creation of a Kuttl test image.  This custom test would execute kuttl to perform testing enabling scorecard users a means to write tests using kuttl but have the result integrated into scorecard results.  

## Design Details

### Custom Test Definition

* A custom test is any sort of test logic required, written in any programming language that satifies the scorecard test requirements for output format and execution.

### Custom Test Components

This proposal outlines the end-user facing components that make up a custom test including:

#### Test Image

A custom test is based on a container image, that image is created by
the end-user using whatever tools they want to use.

The test image must produce log output that conforms to the scorecard Test Result output.

> The test image will be executed as a Pod by the scorecard with a restart policy of `never`.

Reviewer: Did you consider `Job` instead of `Pod`. What are the pros/cons of each?

Author: I did, the only value that I could see from doing this with a Job would be for restarting a test, I didn't see a use case for job restarts with scorecard at this point.  I guess the con's of a job is that its another layer of abstraction to go find the pod log output as well for scorecard and end users alike.   If you can think of a good pro, I'd be for doing this as a Job, but I couldn't come up with one.

Reviewer: That all sounds like good reasoning to use a Pod directly. :+1:

> When executing a scorecard test, scorecard will create a ConfigMap holding the scorecard configuration manifests such as the Custom Resource or Bundle being tested.  This ConfigMap can then be mounted into the custom test image and used within the test execution.  

Reviewer: Consider the use case of running `operator-sdk scorecard --bundle-image=quay.io/myorg/mytest` with no other config file.

For each test execution, we could construct:
1. A ConfigMap containing the required test configuration for the test.
2. A PodSpec with the containers being the bundle image, the test image, and an init container that copies the bundle files from the bundle image into the test image at a specific place (using a shared emptyDir volume mount) (e.g. `/scorecard/bundle`)
3. As you suggest, we would also mount the ConfigMap into the test image container  (e.g. `/scorecard/config`)

I think we should be opinionated for this to decrease the configuration complexity.

> When executing a scorecard test, scorecard will create a ConfigMap holding the scorecard configuration manifests such as the Custom Resource or Bundle being tested.  This ConfigMap can then be mounted into the custom test image and used within the test execution.  

Reviewer: Actually, assuming the bundle contains the scorecard test configuration, we don't necessarily need the separate ConfigMap. Or instead of the init container and the copying, we can have scorecard extract the bundle image and create a ConfigMap from it to be mounted into the test image.

> When executing a scorecard test, scorecard will create a ConfigMap holding the scorecard configuration manifests such as the Custom Resource or Bundle being tested.  This ConfigMap can then be mounted into the custom test image and used within the test execution.  

Author: I've documented the mount point in this latest commit...I also mentioned your bundle image approach, also created a user story to have scorecard work with bundle images in general.

#### Test Output

Custom tests are expected to produce log output that conforms to the
scorecard v1alpha2 Test Result output.  This allows scorecard to parse
the custom test log output and aggregate the results like any other
scorecard test.

> The scorecard expected output would be in JSON, and would match the following definition:  https://github.com/operator-framework/operator-sdk/blob/master/pkg/apis/scorecard/v1alpha2/types.go#L36

Author: I added a JSON snippet from an existing test as an example.

The scorecard v1alpha2 Test Result output contains information such as
pass or fail status and suggestions/errors produced by the test that
the end user would want to see in the aggregated scorecard results.

### Custom Test Development Model

The developer of a custom test would have a workflow similar to:
 
 * create a Pod manifest to run their custom test image, copying a sample Pod manifest from the scorecard examples in the SDK repository

>  * write a custom test binary in the language of their choice, sample custom test within the SDK repository would be available as a starting point

Reviewer 2: Long term note, would be cool to have scorecord api to generate this for different languages. 
```python
import scorecard

scorecard.write_test_result("foobar test", scorecord.PASS)
```
That way we can ensure we get the format we expect and the user doesn't have to keep fumbling to get it right. Again this is a longer term thing not part of this proposal.

Author: editor's note:  I'd LOVE to see a java and/or python test shipped as an example custom test.

 * test the custom test binary locally (out-of-cluster) to make sure the output is produced that matches v1alpha2 ScorecardTestResult format
 * when the test binary works as expected, the developer would build and push the test image to a container registry accessible to the scorecard test environment

>  * the scorecard configuration would be updated to reference the custom test Pod manifest location

Reviewer 2: I vaguely remember why this is needed. I'm okay for now, but it is an annoying step I know I would forget to do.

 * the scorecard would discover the custom test based on the scorecard configuration and also the custom test labels applied by the end-user within the custom test Pod manifest

### Custom Test Packaging and Configuration

In this design proposal, custom tests are Pod manifests in YAML that run a custom test image.

> Note that is would also be possible to add a resource limit to the scorecard configuration as shown in the example below.  Those resource limits would be applied to custom tests that are executed allowing for some control as to how much resources are available for the test execution.

Reviewer 2: typo: "Note that is would" should be "Note that it would"

Author: fixed in latest commit.

> Custom tests (ie Pod manifests) would be accessible via URL or from a local disk location.  Custom test locations would be configured from within the scorecard configuration 

Reviewer 2: typo: ie should be "i.e."

Author: should be fixed in latest commit.

> Custom tests (ie Pod manifests) would be accessible via URL or from a local disk location.  Custom test locations would be configured from within the scorecard configuration 

Reviewer: Do we really need to add the complexity of allowing the pod manifest to be provided via URL?

I think we can assume that the bundle will contain the test configuration and pod manifests.

Author: I like the idea of dropping the 'via URL' feature, will remove that.

```
scorecard:
  # Setting a global scorecard option
  output: text
  tests:
    init-timeout: 60
    resources:
      limits:
        cpu: "1"
        memory: "200Mi"
      requests:
        cpu: 750m
        memory: "100Mi"
      test-path:
        - "tests/here/ondisk"
        - "https://github.com/someisv/tests"
        - "https://github.com/operator-framework/operator-sdk/scorecard-tests"
      cr-manifest:
        - "deploy/crds/cache.example.com_v1alpha1_memcached_cr.yaml"
        - "deploy/crds/cache.example.com_v1alpha1_memcached_cr2.yaml"
  …
```

### Custom Test Execution

Scorecard tests would be executed from the SDK scorecard CLI as is the case today.  Scorecard would execute the tests, capture the results, and provide feedback to the end-user by means of the current scorecard API (e.g. ScorecardTestResult).  Command line flags would allow the end-user a means to specify which tests are executed and upon which test environment to run the tests.  For example:
```

> operator-sdk scorecard -o text --selector=suite=isvcustom

Reviewer 2: :+1:

```

### Custom Test Environments

Scorecard would allow you to run tests on a local Kube cluster as is the case today.  Provisioning the test cluster is external to the scope of the scorecard itself.  For example, some might want to run a scorecard on a provisioned environment such as kind (kube in docker).  

### Custom Test Labeling
As is the case today, custom tests would allow end users to specify custom labels to each test so that the scorecard could select which tests are run based on a labeling scheme.  Here is an example of user added labels to a custom test step:
```
apiVersion: apps/v1
kind: Pod
metadata:
  name: isvtest1
  labels:
    vendor: someisv
    test: isvtest1
    suite: isvcustom
```

The end user could run that test with this command:
```
operator-sdk scorecard -o text --selector=test=isvtest1
```

#### Migrating Scorecard Internal Tests to Custom Test Images

With the above custom test capability, there is benefit to Red Hat in migrating the existing scorecard internal tests to be based on the custom test format. 

> Today there are 2 internal scorecard tests that make use of the scorecard proxy, with these tests migrating to the custom test format, the scorecard proxy is likely not necessary which would also reduce the overall scorecard implementation complexity.

Reviewer 2: :+1: to migrating

### Helm Operator Complex Tests

Writing a custom test image in a Helm chart was not evaluated or pursued in this proposal.  What is likely for a Helm operator is that the operator owner would need to write custom test images using either Python, Java, golang, or some other programming language.

## Alternatives

The alternatives to this design proposal might include:
 * Other existing/open-source test frameworks could be evaluated for use within the SDK for implementing the desired custom test functionality.

## Conclusion

The proposed changes to the scorecard solve the immediate need to support custom or user-provided tests, both simple and complex tests.  However, the longer term implication for scorecard is that all of its tests would evolve to use a single testing format that is far more flexible that what exists today in the current implementation.

The proposed design focuses heavily on separation of concerns, turning scorecard into a test runner essentially, moving test implementations into their own concern (eg. container images).

## Reference Material

> <https://github.com/kudobuilder/kuttl>

Reviewer 2: Could we link to the original proposal as well?

Author: link provided in latest commit.

 * scorecard will determine test configurations by a YAML/DSL file that is either passed on the command line or included in the bundle contents
 * scorecard will only run OLM tests if OLM manifests are found in the bundle
 * scorecard command syntax will look as follows:
   * operator-sdk scorecard --dsl <point-to-yaml> --bundle <bundle-path/image> -l <selector>

### User Stories

#### Story 1

I am as an Operator developer, I'd like to be able to construct custom tests and execute them via the scorecard.

#### Story 2

Scorecard will include an example custom test that end users could use as a reference when writing their custom tests.

#### Story 3

Scorecard internal tests will be converted to the custom test format and therefore be externalized from the scorecard binary itself.

#### Story 4

Scorecard would process tests based on an on-disk directory or bundle image that contains the scorecard configuration.

#### Story 5 

- I am as an Operator developer, would like to be able to create tests using the same Assert syntax adopted by [Kuttl](https://github.com/kudobuilder/kuttl) and check its results aggregated within scorecard results. 

**Implementation Notes**

> The Scorecard would support the creation of a Kuttl test image and the execute tests via Kuttl, however, it would be transparent for users since its results will be integrated into scorecard results.

Reviewer 3: nit: and then execute

Author: the commit just now fixes that type-o, thanks.

## Design Details

### Glossary

 * test image - this is a container image built by end-users to contain a custom test implementation
 * test configuration (DSL) - this is a file that contains the test definitions and labels which determine what tests are run and how they are configured
 * Custom Test Definition - A custom test is any sort of test logic required, written in any programming language that satifies the scorecard test requirements for output format and execution.

#### Test Image

A custom test is based on a container image, that image is created by
the end-user using whatever tools they want to use.

> The test image must produce container log output that conforms to the scorecard Test Result output.  The output is an array of Test Results, so a test image could contain multiple tests if necessary.  Here is a sample of test output in the required format:

Reviewer: ```suggestion
The test image must produce container log output that conforms to the scorecard Test Result output.  The output is a single Test Result, so a single run of a test image should execute and output the result for exactly one test.  Here is a sample of test output in the required format:
```

```
    [
      {
	      "name": "Spec fields with descriptors",
	      "description": "All spec fields have matching descriptors in the CSV",
	      "state": "fail",
	      "suggestions": [
		"Add a spec descriptor for size"
	      ],
	      "crname": "example-memcached"
      }, 
      {
	      "name": "Some other test",
	      "description": "Another test on the CSV",
	      "state": "pass",
	      "suggestions": [
	      ],
	      "crname": "example-memcached"
      },

>     ]

Reviewer: We discussed some changes to this output format, right? Remove the array, include just one test, and remove CR name?

Author: removed array in latest commit.

```

The test image will be executed as a Pod by the scorecard with a restart policy of `never`.

Also, for each test execution, scorecard would also make the
bundle image contents available to test images by mounting
the bundle image contents into the following location inside the test pod:

```

> /scorecard

Reviewer: In case we want to add other directories in the future?
```suggestion
/scorecard/bundle/
```

```

This allows tests to have access to all the bundle contents if they 

> are required for the test logic.  Note:  It is the responsibility of the test to iterate though each CR that they want to test and create Custom Resources as scorecard will no longer be responsible for this.  The list of CRs to test are found in the CSV within the bundle image content.

Reviewer: We're not really making any assumptions about what the test does. It could:
1. Not use the CRs at all (e.g. doing a static bundle validation)
2. Use the CRs in the CSV's `alm-examples` annotation (e.g. to verify they validate the associated CRD schema)
3. Use some other CR(s) built (or maybe mounted) into the test image.

Author: I added this almost verbatim in latest commit.  Thanks!

For the case of executing a test, the bundle image will be passed as a scorecard command flag:
```
operator-sdk scorecard --bundle='quay.io/myorg/mytest' --selector=suite=basic

> ```

Reviewer: What was the decision about `--bundle` vs `--test-assets`, or maybe even just `--mount`? Basically a flag like `--bundle` but not specific to OLM.

I thought we arrived at something more generic so as to not use/require OLM terminology.

We _would_ enable the OLM test configuration if and only if we find the expected OLM bundle directories in the expected location within the mounted docker image/directory.

And we _would_ enable the kuttl test configuration if and only if we find the expected kuttle test directories in the expected location within the mounted docker image/directory.

Author: I was thinking we would have a single flag that would cover both cases a) on-disk directory and b) bundle image, scorecard would look at what is passed to know if the contents are on-disk (exploded) or in an image that needs to be extracted to on-disk.  I'm totally flexible on the name, maybe --test-assets as Joe suggested?  I could also make it 2 flags is that is a better end user experience.

Reviewer: Part of the plan is to drastically simplify config/flag sprawl from the current scorecard. My vote is that if we want to keep `--bundle` lets just use that for now (and it can accept a bundle image or bundle directory).

Then if we want to become more generic later, we can add a new flag and optionally deprecate `--bundle` if the new flag incorporates the features of `--bundle`. WDYT?

#### Test Output

Custom tests are expected to produce log output that conforms to the
scorecard v1alpha2 Test Result output.  This allows scorecard to parse
the custom test log output and aggregate the results like any other
scorecard test.

The scorecard expected output would be in JSON, and would match the following definition:  https://github.com/operator-framework/operator-sdk/blob/master/pkg/apis/scorecard/v1alpha2/types.go#L36

The scorecard v1alpha2 Test Result output contains information such as
pass or fail status and suggestions/errors produced by the test that
the end user would want to see in the aggregated scorecard results.

> ### Custom Test Development Model

Reviewer 4: This feels like a very long list of steps and might not convince a lot of developers to actually ship tests.

How about the following being the default for most users, especially when they start:

1. Developer deploys kuttl-based assertion tests in the form of YAML files in a known directory location
2. SDK has a `test` command which
   - automates the creation of the scorecard config referencing the kuttl tests
   - runs the tests using the scorecard configuration on disk with no particular label selector (all tests)
   - supports a `--namespace` tag, should the Operator prefer a certain watch namespace
3. when the user is happy the SDK has a `build` command which builds the bundle with the test definition and scorecard configuration from 2)

Reviewer: @jmccormick2001 @shawn-hurley correct me if I'm wrong but I think the idea is that if an operator developers runs `operator-sdk scorecard` with no options, the following would happen:
1. Scorecard would initialize an empty list of tests 
2.  Scorecard would append the basic test configuration to the list of tests
3. If a bundle directory exists at a known project location, scorecard would append the OLM test configuration to the list of tests
4. If a kuttl directory exists at a known project location, scorecard would append the kuttl test configuration to the list of tests.
5. If a DSL yaml file exists at a known project location, scorecard would append the tests configured in the DSL file to the list of tests.
6. Scorecard runs all of the tests that were setup in steps 2-5.

To package the bundle, the operator would run `operator-sdk bundle create` as usual. We would  add the DSL file and kuttl directory to the bundle if we find them.

I'm thinking we add a `scorecard` directory to the project and store the DSL config file and kuttl directories there. Then the bundle dockerfile just needs `COPY scorecard /scorecard`

Author: yep, 
 * we would need the 'operator-sdk bundle create' to be able to do the image build
 * we would run the default basic/olm tests if no selector is added (like we it works today)
 * we need a place to store the scorecard assets (config and other), a 'scorecard' subdir under the project would be fine with me
 * I was thinking that when a person builds a new project it would scaffold out the scorecard subdir with the basic/olm tests being configured by default, I suppose you could also have a 'scorecard --scaffold' command that users could run afterwards to re-scaffold based on what is in the scorecard subdir.

if this works for everyone I could update the proposal to spell it out further.

Reviewer 3: let's talk about the name of the test dir in another followup enhancement specifically in operator-framework/enhancements.

I agree that this is the correct default behavior of the scorecard test.

Reviewer: I still have a question about the basic and olm tests. Do we expect to find them explicitly configured in the DSL file, or do we build their configs into the binary and inject them automatically?

What would the scaffolded `scorecard/config.yaml` file look like after `operator-sdk new` (or `init` when we switch to KB's CLI for scaffolding)?

Reviewer: If they're explicitly in the file, that gives users the opportunity to permanently remove them without using selectors.

Author: I was planning on making them explicit in the config file to allow users to tweak as they want, however, I see the case for just keeping them inside the scorecard by default if we want to force them to always be available.

Author: perhaps if the goal with this is to always have a known set of tests (olm/basic) that CVP can assume will always be there, then you'd have to go with the approach of keeping them inside scorecard itself instead of a config file.

Reviewer: I would be fine with either method. If we go the explicit route, then the `kuttl` tests should be explicit as well for consistency.

For the pipeline/CI case, what happens if a user can't get the basic tests to pass, so they just remove them from the config file?  Do we envision the pipeline injecting it's own test configs? 

How would that work if the DSL is built into the bundle image and scorecard doesn't give the pipeline a chance to inject anything since it handles the bundle image extraction internally?

Reviewer: We might need to answer that question regardless, if there are other policy tests NOT built into scorecard that a pipeline would want to enforce.

Reviewer 3: because they are optional based the fact that a directory exists, I don't think they can be in the config file right?

Author: here is what I'm currently thinking:  if a pipeline requires tests to be run, then they would require those tests be added to the config file, if the tests are not in the config file then the pipeline fails the test...kuttl is the only case where the tests that actually are run would be driven by what kuttl tests are found in the kuttl specific directory.  We could start with explicit tests in the config and later address injecting those behind the scenes if it turns out we need to do that for pipeline reasons.

The developer of a custom test would have a workflow similar to:
 

>  * write a custom test binary in the language of their choice, sample custom test within the SDK repository would be available as a starting point

Author: I've added a section on the kuttl test case, also updating the developer workflow.

 * test the custom test binary locally (out-of-cluster) to make sure the output is produced that matches v1alpha2 ScorecardTestResult format
 * create a scorecard DSL yaml file to run their custom test image
 * when the test binary works as expected, the developer would build and push the test image to a container registry accessible to the scorecard test environment

>  * the developer would rebuild their bundle image passing in their scorecard DSL file

Reviewer: I think the default would be empty since the Basic, OLM, and Kuttl tests would be automatically configured without the presence of the file.

>  * the developer would rebuild their bundle image passing in their scorecard DSL file

Author: I've updated this section to allude to the scaffolding potential, especially in generating the scorecard DSL config file and building the bundle (test) image.

> The case of a developer implementing a kuttl test would be as follows:

Reviewer 3: the default kuttl, and olm test images must be overridable IMO.

Author: I added a sample of the DSL that shows how default images could be overridden.

 * the Basic and OLM test manifests would be added automatically to the bundle image as part of the bundle image build process.  This allows users to run the Basic and OLM tests along with their custom tests.
 * lastly, the developer would run the scorecard passing in their operator bundle image they just built along with any other scorecard flags (e.g. selector).

### Custom Test Packaging and Configuration

> In this design proposal, custom tests are container images, described in the scorecard DSL yaml file.  End users would add their scorecard DSL yaml file into their operator bundle image or pass it into scorecard using a command line flag.  Custom test images are stored in a container registry outside the control of scorecard.

Reviewer: If the DSL config file is found in both the expected bundle (or test asset/mount) image/directory AND from the flag, does the flag take precedence?

Author: good catch, yes, I'd think the command line flag would take precedence.

Author: updated in latest commit.

### Scorecard Configuration

The scorecard DSL yaml file would be accessible from a local disk location to scorecard or from within a bundle image. 

Scorecard would assume that the scorecard DSL yaml file would be located in the bundle image at the following location:
```

> /scorecard/.osdk-scorecard.yaml

Reviewer: Let's take an opportunity to rename this to a non-hidden file.

How about simply `/scorecard/config.yaml`?

Author: updated in latest commit.

### User Stories

#### Story 1

I am as an Operator developer, I'd like to be able to construct custom tests and execute them via the scorecard.

#### Story 2

Scorecard will ship with an example custom test that end users could use as a reference when writing their custom tests.

#### Story 3

As a scorecard developer I would like for custom tests and internal scorecard tests to share the same architecture to provide a cleaner scorecard design and implementation. Scorecard internal tests will migrate to the custom test format and therefore be externalized from the scorecard binary itself.

#### Story 4

I am as an Operator developer, would like to be able to use the Scorecard feature to test the [OLM Bundles](https://github.com/operator-framework/operator-registry/tree/v1.5.3#manifest-format) created for my project.

#### Story 5 (optional)

Scorecard would support the creation of a Kuttl test image.  This custom test would execute kuttl to perform testing enabling scorecard users a means to write tests using kuttl but have the result integrated into scorecard results.  

> ## Design Details

Reviewer: Somewhere in here it would be nice to walkthrough a full example of a typical user interaction:
1. The user has their operator and bundle images pushed to a container registry.
2. Let's assume the Basic and OLM tests exist as separate test images. What does an operator developer do to run their operator against these tests?
3. Now let's assume the same bundle is being run in a pipeline where there is no source code available to the pipeline. How is the same exact test scenario executed in that case?
4. For 2 and 3, what is the step-by-step process that is happening under the hood in the scorecard code?

Author: I took a shot at that workflow, based on the bundle-image approach in this latest commit.

Author: I didn't call it out yet in the proposal, but I suspect we want to just make having OLM installed a mandatory requirement going forward, and deploy the operator using OLM?  The scorecard could test whether OLM is installed I suspect and present the user an error message OLM is not found with a note on how to install OLM.  Agree?  If so, this simplifies scorecard somewhat in that it doesn't have to support non-OLM environments as well as OLM environments any longer.

      },
    ]
```

The test image will be executed as a Pod by the scorecard with a restart policy of `never`.

When executing a scorecard test, scorecard will create a ConfigMap holding the scorecard configuration manifests such as the Custom Resource or Bundle being tested.  This ConfigMap can then be mounted into the custom test image and used within the test execution.  

The ConfigMap would be mounted into the test container at the following location:
```
/scorecard/config
```

For the case of testing a bundle image, the following command will run scorecard tests which are configured within the bundle image itself:
```
operator-sdk scorecard --bundle-image=quay.io/myorg/mytest
```

In this case, for each test execution, we could construct:

 * A ConfigMap containing the required test configuration for the test.
 * A PodSpec with the containers being the bundle image, the test image, and an init container that copies the bundle files from the bundle image into the test image at a specific place (using a shared emptyDir volume mount) (e.g. /scorecard/bundle)
 * We would also mount the ConfigMap into the test image container (e.g. /scorecard/config)

> For this, scorecard would need to support bundle images which is included as a User Story in this proposal.  If the bundle image were to contain the scorecard configuration it might be possble to use that for running the scorecard test as opposed to having a scorecard configuration outside of the bundle image being used.

Reviewer: I think we should talk through this a little more.

Can the bundle image be the one and only required command line flag that dictates where to find the configuration for what to test? Put another way, are there any other mechanisms that we want (or need) to support to enable scorecard to know what to test (e.g. local filesystem directory). 

IMO, I think we start out assuming that the bundle image will contain all the necessary info and configs to run the full suite.

There are three classes of configuration in my mind:
1. Configuring `operator-sdk scorecard` to know what test images to run.
2. Configuring each test image with the bundle contents.
3. Configuring each test image with its custom configuration.

If the assumption is that the bundle image actually contains the first class of configuration, that means that scorecard will need to be able to fetch and extract the contents of the bundle image locally so that it can know what test images to run in the cluster. This could be a problem in the CI/pipeline case where access to a Docker socket is not available. Are there other OCI tools or libraries available that can pull and extract images without requiring root?

Then we have the question about the method of providing configuration to the test images. It looks like there are a couple of different methods listed here. Do we have enough information now to choose one of those?

**Method 1** Add bundle image container as sidecar, use initContainer to copy bundle files into emptyDir shared with the test container.

This method is desirable because it doesn't require Docker (or any other OCI utilities) to exist locally. This seems like a big benefit for the CI/pipeline use case, but maybe we need those tools locally anyway if we want scorecard to get its test configuration from the bundle itself?

**Method 2** Scorecard pulls and extracts the bundle image locally, creates a ConfigMap of the contents, and launches the test container with the ConfigMap mounted in.

This method is more straightforward on the PodSpec side since it just involves creating a ConfigMap and it in to the test container.

Author: I wonder if people will not like having their test assets (scorecard config + test manifests) stored within something like a operator bundle?   Some places I've worked at would not want to ship tests within a production bundle for a variety of reasons.  But if thats not an issue, then Method 1 seems nice because it keeps you from having test assets laying about on local disk or ConfigMaps being created and then having to be cleaned up.

Author: I suppose for a production bundle, you could just strip out the scorecard assets as some sort of workflow or CLI command.

Reviewer: @shawn-hurley @dmesser Thoughts?

Reviewer 4: I think we should support both scenarios - test configuration shipped in parallel with the bundle and as part of the bundle. The latter is just making everything very portable and simplifies the UX. Like @joelanford said, it should be possible to just give scorecard the bundle and it knows, what to do.
It should be noted that the Operator bundle is not the actual production application itself. Pods running on cluster will not have access to any test scripts or manifests. See [bundle extraction on cluster](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/contributors/design-proposals/pull-bundle-on-a-cluster.md) for more info.
There is precedence for shipping test configurations with the application packaging with `helm`.

If some pipeline / organization absolutely cannot deal with test scripts being part of the bundle, there should be a second option to instruct scorecard where to find the data. The tradeoff for them is that it makes those people jobs harder to orchestrate a test run and find the right version of the test configuration for the bundle at hand.

Reviewer: Seems reasonable. How about this:

* Add a `--test-asset-image` flag.
* The test asset image must have the exact same filesystem layout for the scorecard tests as the bundle image.
* If `--test-asset-image` is provided, scorecard get all configs from that image and none from the bundle image. This covers the case that the bundle image ALSO has test assets in it.

 * write a custom test binary in the language of their choice, sample custom test within the SDK repository would be available as a starting point
 * test the custom test binary locally (out-of-cluster) to make sure the output is produced that matches v1alpha2 ScorecardTestResult format
 * when the test binary works as expected, the developer would build and push the test image to a container registry accessible to the scorecard test environment
 * the scorecard configuration would be updated to reference the custom test Pod manifest location
 * the scorecard would discover the custom test based on the scorecard configuration and also the custom test labels applied by the end-user within the custom test Pod manifest

### Custom Test Packaging and Configuration

In this design proposal, custom tests are Pod manifests in YAML that run a custom test image.

Note that it would also be possible to add a resource limit to the scorecard configuration as shown in the example below.  Those resource limits would be applied to custom tests that are executed allowing for some control as to how much resources are available for the test execution.

> Custom tests (i.e. Pod manifests) would be accessible from a local disk location.  Custom test locations would be configured from within the scorecard configuration 

Reviewer: Would this work in the pipeline case where source code is not available?

Author: I was originally going to have the scorecard test pod manifests accessible via URL to avoid CI/pipelines from having to have a local copy or have to clone them, but if they are stored in the bundle itself, then that would negate having to have them locally or pull them down via URL.  So, I think making the decision to store scorecard config + test pod manifests inside the bundle image would solve this problem for the CI/pipeline use case.

file.  Custom test labels would be also in the scorecard configuration file whereas the scorecard could pick tests to run by test labels.  For example, the scorecard configuration might be as follows to indicate the presence of custom tests from different sources:
```
scorecard:
  # Setting a global scorecard option
  output: text
  tests:
    init-timeout: 60
    resources:
      limits:
        cpu: "1"
        memory: "200Mi"
      requests:
        cpu: 750m
        memory: "100Mi"
      test-path:
        - "tests/here/ondisk"
        - "someisv/tests"
        - "operator-sdk/scorecard-tests"
      cr-manifest:
        - "deploy/crds/cache.example.com_v1alpha1_memcached_cr.yaml"

>         - "deploy/crds/cache.example.com_v1alpha1_memcached_cr2.yaml"

Reviewer: Is it reasonable to remove the `cr-manifest` configuration and require the CR manifests under test to be included in the CSV examples?

Reviewer: [This proposal](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/contributors/design-proposals/internal-objects.md#hiding-internal-concepts-from-end-users) is also relevant. If a CRD is listed as data-only, we should not make the same assumptions that we would about workload CRDs.

I'm not sure how prevalent usage of these annotations are yet, so maybe something that can be considered in a follow-on.

Author: Seems reasonable to have the scorecard drop cr-manifest from its scorecard config and instead just pull them from the CSV examples.  Probably this works is done with the 'make bundle mandatory' story.  Same applies for the scorecard config's csv-manifest, that would no longer be used, but instead the CSV within the bundle.

The annotation proposal is interesting, I think its something we would need to address when that feature is available based on my reading of that.

---
title: Scorecard Custom Test Framework
authors:
  - "@jemccorm"
reviewers:
  - TBD
  - "@joelandford"
  - "@zeus"
approvers:
  - TBD
  - "@shurley"
  - "@dmesser"
creation-date: 2020-03-19
last-updated: 2020-03-19
status: implementable
---

# Scorecard - Custom Tests

## Release Signoff Checklist

> - \[ \] Enhancement is `implementable`

Reviewer 5: ```suggestion
- \[ x\] Enhancement is `implementable`
```

## Motivation

The scorecard needs a robust means of enabling end-users of the SDK to write and execute their own operator tests.   Ideally, the custom test framework would decouple test logic from within the SDK itself, making the SDK and scorecard logic in particular more flexible and maintainable in the long run.

## Goals

 * Document a design for scorecard to execute end-user or custom tests.

 * Document stories/steps required to deliver scorecard custom test functionality.

### Non-Goals

 * This is a design proposal and does not include a proof-of-concept.

## Proposal

This design proposal is to establish a pattern for implementing custom scorecard tests and also support the goal of migrating existing scorecard tests into the same format as the custom tests.

With this proposal, the scorecard will continue to setup and teardown the Operator being tested, as well as creating Custom Resources being tested.

### User Stories

#### Story 1

> Scorecard users will be able to construct custom tests and execute them using the scorecard.

Reviewer 5: WDYT about writing the story with the roles papers?
IMO it makes more clear the understatement since we know who wants/will use and for o what.
E.g

```suggestion
I am as an Operator developer, I'd like to be able to construct custom tests and execute them via the scorecard.
```

#### Story 2

> Scorecard will ship with an example custom test that end users could use as a reference when writing their custom tests.

Reviewer 5: WDYT?
```suggestion
I am as an Operator developer, I'd like to be able to write my custom tests from examples scaffolded via scorecard.
```

Author: hmmn, I wasn't thinking of scaffolding test examples with this initially but rather just include some examples already in working form.

Reviewer: I agree with Jeff. Let's keep the scope of this as narrow as possible so we focus on the scorecard execution and interfaces.

Reviewer: @jmccormick2001 I think a good test example would be a test image that runs `operator-sdk bundle validate` against the bundle that is mounted into the test image.

It's a simple test scenario that has very few moving parts, the code pretty much already exists for it, and it is actually useful and will likely be the first official test image used in CI and pipelines.

#### Story 3

> Scorecard internal tests will migrate to the custom test format and therefore be externalized from the scorecard binary itself.

Reviewer 5: Why it is important for the dev? What is the goal with it? I mean, what will be possible to achieve with?

Author: I'll make a note about this, thx.

#### Story 4

> Scorecard would support testing a bundle image.  

Reviewer 5: ```suggestion
I am as an Operator developer, would like to be able to use the Scorecard feature to test the [OLM Bundles](https://github.com/operator-framework/operator-registry/tree/v1.5.3#manifest-format) created for my project.
```

#### Story 5 (optional)

> Scorecard would support the creation of a Kuttl test image.  This custom test would execute kuttl to perform testing enabling scorecard users a means to write tests using kuttl but have the result integrated into scorecard results.  

Reviewer 5: ```suggestion
- I am as an Operator developer, would like to be able to create tests using the same Assert syntax adopted by [Kuttl](https://github.com/kudobuilder/kuttl) and check its results aggregated within scorecard results. 
```

## Design Details

> 

Reviewer 5: WDYT about before described the details over each smaller unit, do a technical overview of how it actually will work to make easier the understatement?  

Something such as: 

```suggestion
 A custom test is any sort of test logic required, written in any programming language that satisfies the scorecard test requirements for output format and execution.
 
In this way, the Operator developer is responsable to provide an image with all requirements and instruction to executed their tests. Scorecard requires that some standards are followed such as the configuration that will be used by the feature and the format of its output logs which will be parsed and integrated to the Scorecard results in order to work.  

The scorecard usage will be like:

`operator-sdk scorecard --bundle-image=quay.io/myorg/mytest`

In this way, scorecard basically will run the project via its bundle and the custom tests image provided and then, get its output logs to be used in its own results. 

**NOTES** See that SDK already has a feature to create the [bundle](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/olm-catalog/operator-bundles.md). 
```

PS: If I properly understood too :-P

Author: I updated that command a bit to make it clearer (I hope) :smile:

### Custom Test Definition

* A custom test is any sort of test logic required, written in any programming language that satifies the scorecard test requirements for output format and execution.

### Custom Test Components

This proposal outlines the end-user facing components that make up a custom test including:

#### Test Image

A custom test is based on a container image, that image is created by
the end-user using whatever tools they want to use.

> The test image must produce log output that conforms to the scorecard Test Result output.  Here is a sample of test output in the required format:

Reviewer 5: Where these logs will be outputted? Would it be in some configurable path?

Author: I'll add a note that this output is in the test pod container log.

```
    {
      "name": "Spec fields with descriptors",
      "description": "All spec fields have matching descriptors in the CSV",
      "labels": {
        "necessity": "required",
        "suite": "olm",
        "test": "specdescriptorstest"

>       },

Reviewer: If the `Pod` YAML is the source of truth for the test labels, does this mean the pod would need to introspect itself to be able to output its own labels?

Or could we just say that the pod output doesn't have to contain the labels since we already know them from the Pod YAML?

Reviewer: Adding to that question.. Can a single execution of a pod involve multiple tests? Or is this a 1 pod execution maps to 1 test kind of thing?

Author: hmmn, I was thinking 1 test per pod execution with 1 result, however, I can see some good value in having the ability to return n-number of test results to simplify test image mgmt.  It shouldn't be hard to support a list of test results coming back from the test.  I'll add that into the next commit of this proposal.

The pod output would not need to contain the labels, the Pod manifest would hold the labels so the results don't have to include them.  The test selection would be driven off of the pod manifest labeling.

### Custom Test Labeling
As is the case today, custom tests would allow end users to specify custom labels to each test so that the scorecard could select which tests are run based on a labeling scheme.  Here is an example of user added labels to a custom test step:
```
apiVersion: apps/v1
kind: Pod
metadata:
  name: isvtest1
  labels:
    vendor: someisv
    test: isvtest1
    suite: isvcustom
```

The end user could run that test with this command:
```
operator-sdk scorecard -o text --selector=test=isvtest1
```

#### Migrating Scorecard Internal Tests to Custom Test Images

With the above custom test capability, there is benefit to Red Hat in migrating the existing scorecard internal tests to be based on the custom test format. 

Today there are 2 internal scorecard tests that make use of the scorecard proxy, with these tests migrating to the custom test format, the scorecard proxy is likely not necessary which would also reduce the overall scorecard implementation complexity.

### Test Plan

> **Note:** *Section not required until targeted at a release.*

Reviewer 5: I think we can remove the section so. WDYT?

Author: agree, will remove in next commit.

### Upgrade / Downgrade Strategy

To implement this custom test functionality, additions or changes to the existing scorecard configuration file and command line flags would be necessary and would not be supported in previous scorecard releases.

## Implementation History

 * Proposal initiated in March 2020.

> 

Reviewer 5: Is this info required? The date of the proposal is in the header already.

Author: will remove in next commit.

their operator bundle images.

For this, scorecard would need to support bundle images which is included as a User Story in this proposal.  

#### Test Output

Custom tests are expected to produce log output that conforms to the
scorecard v1alpha2 Test Result output.  This allows scorecard to parse
the custom test log output and aggregate the results like any other
scorecard test.

The scorecard expected output would be in JSON, and would match the following definition:  https://github.com/operator-framework/operator-sdk/blob/master/pkg/apis/scorecard/v1alpha2/types.go#L36

The scorecard v1alpha2 Test Result output contains information such as
pass or fail status and suggestions/errors produced by the test that
the end user would want to see in the aggregated scorecard results.

### Custom Test Development Model

The developer of a custom test would have a workflow similar to:
 
 * write a custom test binary in the language of their choice, sample custom test within the SDK repository would be available as a starting point
 * test the custom test binary locally (out-of-cluster) to make sure the output is produced that matches v1alpha2 ScorecardTestResult format

>  * create a Pod manifest to run their custom test image, copying a sample Pod manifest from the scorecard examples in the SDK repository

Reviewer 6: It would be nice to have a script/set of flags in `scorecard` to generate a basic Pod manifest for a test, with some `# Add <thing> here` comments in various fields that we can't fill in via CLI. Not necessarily something that needs codifying in this document, just a thought for a followup.

Author: good idea, I was thinking either good examples people could copy but I see value in some scaffolding to generate a test based on a given operator's configuration/naming.

### Custom Test Components

This proposal outlines the end-user facing components that make up a custom test including:

#### Test Image

A custom test is based on a container image, that image is created by
the end-user using whatever tools they want to use.

The test image must produce container log output that conforms to the scorecard Test Result output.  The output is an array of Test Results, so a test image could contain multiple tests if necessary.  Here is a sample of test output in the required format:
```
    [
      {
	      "name": "Spec fields with descriptors",
	      "description": "All spec fields have matching descriptors in the CSV",
	      "state": "fail",
	      "suggestions": [
		"Add a spec descriptor for size"
	      ],
	      "crname": "example-memcached"
      }, 
      {
	      "name": "Some other test",
	      "description": "Another test on the CSV",
	      "state": "pass",
	      "suggestions": [
	      ],
	      "crname": "example-memcached"
      },
    ]
```

The test image will be executed as a Pod by the scorecard with a restart policy of `never`.

> When executing a scorecard test, scorecard will create an environment variable in the test container that holds the name of the CR being tested, this allows the test writer a means of knowing which CR is being created by the scorecard as part of this test execution.

Reviewer: My initial instinct here is that the test image would be responsible for doing whatever setup is necessary from the files it cares about from the bundle.

So if a scorecard test wants to deploy a CR and then make some assertions about what happens as a result, great! But if a scorecard test wants to do a static validation of the bundle, then there's no need to create a CR for that test.

Moreover, I think creating one or more CRs is a step that test images _must_ do because some custom tests may require a complex setup of CRs. For example, this test scenario should be possible: 
1. Create a data CR with some configuration about DB credentials
2. Create an app CR to run an instance of a DB
3. Check that the DB is healthy
4. Populate some data in the DB
5. Create a CR that triggers a backup of the DB
6. Check that the backup contains the populated data.  

None of this is possible with a simplistic single CR test scenario, so I think we leave the test setup up to the test image. All scorecard would do at that point is fetch the bundle, extract the configuration from the bundle about what tests to run, make the bundle contents available to the test containers, run each of the tests, and capture and report their results.

Author: sounds reasonable to remove the CR creation from scorecard (the way it works today), and instead go with this approach, that being to let the tests iterate thru the CRs, creating the CRs as they wish.  I suppose the scorecard might still need to clean up the CRs after a test execution still?  or do you think the tests should honor cleaning up after themselves as part of their logic and we don't assign that to the scorecard at all?

Reviewer: I think scorecard will likely create a namespace to run the pod for the test image and will delete that namespace after the test completes. So anything in that namespace would get automatically cleaned up, but I think the test image should do its own setup/cleanup if it needs anything more than that.

Author: ah ok, I'd not counted on scorcard handling namespace creation/deletion, but if that feature is present it simplifies cleanup of test resources for sure.  Probably a new story for handling the namespaces is necessary,  I'll add it in the next commit.

Author: I updated the scorecard processing/workflow section to show it creating and deleting namespaces as it interates thru the tests, removing the CR creation.

Reviewer: Do you think it makes sense for scorecard to use the namespace from the environment (e.g. `KUBECONFIG` or `~/.kube/config`, etc.)?

I can see an argument for that too. That would simplify the scorecard execution side of things.

Another thing I just thought of is that test containers will need a service account that gives them enough privileges to execute the test. We need to figure out how that should work.

Reviewer 5: Should not it work as operator-sdk run --local?
Then, should not it have an --operator-namespace and --watch-namespace flags?
Otherwise how would work the operator cluster-scope scenario? @joelanford

Author: for service account's in the test pods, we could either a) use the operator's service account which would probably be more privs than necessary, or b) look for a role/binding/SA using a naming convention perhaps (e.g. test-role.yaml, test-role-binding.yaml, etc.) and then create the SA/role/binding on the fly before the tests are run.  Probably option 'b' is suitable since it means the test writer is responsible for creating exactly the SA/role that their tests need, scorecard would just create them as necessary.  For option 'b', either a naming convention to find the security manifests or a CLI flag/config option would tell scorecard where those are to be found.

Not having scorecard deal with namespace creation/deletion greatly simplifies the scorecard implementation.  I'm not sure how scorecard would deal with setting up n-number of namespaces based on varying operator configurations, that could get tricky to figure out and setup correctly.  I'd probably narrow down what scorecard could do here, that would be to work with a single namespace, one currently created, and one that both the operator and the tests could run in at the same time.  Cleanup of resources would be within that namespace instead of depending on namespace deletion for cleanup.

Reviewer: What if we went with:
1. We deploy the test images in the same namespace as the operator
2. We use the same service account that the operator is using

It stands to reason that the operator's service account should be pretty close to the superset of rules that a test needs. If that assumption doesn't hold, we can add more flexibility for custom roles later?

Author: :+1:

### Custom Test Packaging and Configuration

In this design proposal, custom tests are Pod manifests in YAML that run a custom test image.  End users would add their test pod manifests and scorecard configuration file into their operator bundle image using a CLI command developed as part of this proposal.

### Scorecard Configuration

Note that it would also be possible to add a resource limit to the scorecard configuration as shown in the example below.  Those resource limits would be applied to custom tests that are executed allowing for some control as to how much resources are available for the test execution.

Custom test Pod manifests would be accessible from a local disk location to scorecard.  Scorecard would obtain the custom test manifests as part of its bundle image processing since the bundle image is passed into scorecard as a configuration flag.  

Scorecard would assume that the scorecard configuration file and test
manifests would be located in the bundle image at the following location:
```
/bundle/scorecard/.osdk-scorecard.yaml
/bundle/scorecard/tests

> ```

Reviewer: We need to discuss the specific bundle locations with @ecordell

Author: so, thinking out loud here...

you might have a command like this:

operator-sdk  bundle  create --directory ./deploy/olm-catalog/memcached-operator --channels stable,beta --default-channel stable quay.io/example/test-operator:v0.1.0 **--scorecard ./scorecard** 

so, adding a flag to the above command would create a bundle image that also pulls in the scorecard config/manifests from local disk location (e.g. ./scorecard)...

to be consistent with the other bundle image LABELS you could have a LABEL like maybe:
LABEL operators.operatorframework.io.bundle.scorecard.v1=scorecard/

Author: https://github.com/operator-framework/operator-registry/blob/master/pkg/lib/bundle/build.go#L53

or maybe its easier given the above impl to just store the scorecard stuff under the existing manifest directory as passed in with "--directory", if that will work, then no changes to operator-registry would be required to have 'other' directories.  With this approach, I'm assuming that adding scorecard manifests into this directory won't cause downstream problems.

Reviewer 6: We should at least default to a well-known location within the image, perhaps enforce a specific location. I don't think configuring location within the bundle image adds much positive UX.

Reviewer: I think the location _within_ the image should be defined by the bundle spec and enforced. That's what I was referring to about getting @ecordell's input.

The location in the project directory could be `<project-root>/test/scorecard` (or whatever would be a sibling of the e2e tests).

Test Pod manifests would be configured from within the scorecard configuration file.  The following is a scorecard configuration file example using this scheme:

```
scorecard:
  # Setting a global scorecard option
  output: text
  tests:
    init-timeout: 60
    resources:
      limits:
        cpu: "1"
        memory: "200Mi"
      requests:
        cpu: 750m

>         memory: "100Mi"

Reviewer: Since the test definitions are pod manifests that support embedding the resources directly, do we need this to be defined and managed by scorecard?

Author: the only reason to define them at the scorecard configuration level itself is if you want to enforce those particular limits across all tests, but problems might arise with this plan if you had tests that had to have different resource limits in order to run, so this might be adding extra complexity than necessary for an initial implementation.

Reviewer: My vote would be to rely on pods to define their own limits. If it becomes apparent that global resource definitions are necessary, we can add it later.

Author: I removed the resources concept in this latest commit.

```
operator-sdk scorecard -o text --selector=suite=isvcustom --bundle-image='some/bundle/image:v0.0.1'
```

Running the above command would have scorecard perform the following:
 
 * fetch the test configuration and test manifests from the bundle image
 * select the tests to run based on the test manifest labels and the selector
 * iterate through each CR within the bundle image CSV
 * execute the test image Pod for each selected test passing in the current CR as an environment variable
 * fetch the test image Pods log output for aggregation of test results
 * display the test results 

### Custom Test Environments

Scorecard would allow you to run tests on a local Kube cluster as is the case today.  Provisioning the test cluster is external to the scope of the scorecard itself.  

### Custom Test Labeling

> As is the case today, custom tests would allow end users to specify custom labels to each test so that the scorecard could select which tests are run based on a labeling scheme.  Here is an example of user added labels to a custom test step:

Reviewer: I'm going to make some statements that will probably result in more discussion and thought. Tell me what you think.

1. Tests are defined as Pods
2. Each pod that runs can return multiple test results.
3. Scorecard test results have names, descriptions, and labels.
4. Pod `metadata.labels` correspond to the scorecard test labels for the purposes of selection. So labels apply at the pod level to all of the tests returned by the image when executed.
5. Scorecard uses the pod `name` when displaying tests selected by the labels with `--list`
6. When the pod runs, it returns multiple results, which will each have different names defined by the test image. 
7. Scorecard uses something in the pod (maybe an annotation?) to display test descriptions selected by labels with `--list`

I'm getting a little hung up on the issue of having two layers where names, labels, and descriptions are relevant. I think we need more specifics around this.

Would it cause too much cluster and configuration churn if we said that a pod can return only one test result?

Author: yeah, the labeling breaks down if we support multiple tests in a single pod spec, to make it work you would need to implement a set-based label expression "test in (test1,test2)", but then you would also need to implement a richer configuration to have multiple test descriptions and test lables, it turns into a mess.

So yes, I think we just go with 1 pod = 1 test to simplify for both the scorecard and for the end user as well.  So that would be my vote.  

The test description I think is just another metadata.label if we go with a single test per Pod approach.

Reviewer: I think label values have a max length of 63 characters (unless that's been changed recently?). If that's the case, that's why I suggested an annotation for description.

Author: good point.

 * if the test results meet their goals, then scorecard could build a bundle image that contains the kuttl tests in a distributable format (e.g. bundle image)

### Custom Test Packaging and Configuration

In this design proposal, custom tests are container images, described in the scorecard DSL yaml file.  End users would add their scorecard DSL yaml file into their operator bundle image or pass it into scorecard using a command line flag.  Custom test images are stored in a container registry outside the control of scorecard.  The scorecard DSL command flag would take precedence over the scorecard DSL config file found in the bundle image if specified.

### Scorecard Configuration

The scorecard DSL yaml file would be accessible from a local disk location to scorecard or from within a bundle image. 

Scorecard would assume that the scorecard DSL yaml file would be located in the bundle image at the following location:
```
/scorecard/config.yaml
```

The scorecard DSL format is yet to be finalized but it would need to include configuration settings for each test including labels.  Here is a sample of what the scorecard DSL configuration would look like:
```

> tests:

Reviewer: if we want to add `parallel: true` later, I think we need something like:
```
stages:
- parallel: false
  tests:
  - image: image1
    description: image1 test
  - image: image2
    description: image2 test
- parallel: true
  tests: [ ... ]
- parallel: false
  tests: [ ... ]
```

# here is an example of a pair of custom tests that an ISV might implement

> - imagename: quay.io/someuser/customtest1:v0.0.1

Reviewer 3: nit: image not imagename

Author: fixed in latest commit.

  description: some ISV test
  labels:
    suite: custom
    test: customtest1
- imagename: quay.io/someuser/customtest2:v0.0.1
  description: some ISV test
  labels:
    test: customtest2
# here is an example of the scorecard basic tests
- imagename: quay.io/operator-framework/scorecard-basictests:v0.0.1
  description: redhat - check spec
  labels:
    suite: basic
    test: basic-check-spec-test
- imagename: quay.io/operator-framework/scorecard-basictests:v0.0.1

>   description: redhat - check status

Reviewer: Since multiple tests can use the same image, we need to expose `entrypoint`.

Author: good catch, I added that in this latest commit.

- imagename: quay.io/operator-framework/scorecard-olmtests:v0.0.1
  description: redhat - CRDs have validation
  labels:
    suite: olm
    test: olm-crds-have-validation-test
- imagename: quay.io/operator-framework/scorecard-olmtests:v0.0.1
  description: redhat - CRDs have resources
  labels:
    suite: olm
    test: olm-crds-have-resources-test
- imagename: quay.io/operator-framework/scorecard-olmtests:v0.0.1
  description: redhat - spec descriptors
  labels:
    suite: olm
    test: olm-spec-descriptors-test
- imagename: quay.io/operator-framework/scorecard-olmtests:v0.0.1
  description: redhat - status descriptors
  labels:
    suite: olm
    test: olm-status-descriptors-test
- imagename: quay.io/operator-framework/scorecard-olmtests:v0.0.1
  description: redhat - bundle validation
  labels:
    suite: olm
    test: olm-bundle-validation-test
# here is an example of the scorecard kuttl tests
- imagename: quay.io/operator-framework/scorecard-kuttltests:v0.0.1
  description: redhat - kuttl tests
  labels:
    suite: kuttl
```

Users could override the default test images shipped with scorecard
by adding the following into the scorecard DSL configuration:
```

> default-images: 

Reviewer 3: I don't understand this section, if we are forcing everything in the config file.

I think CLI options to override the images are also something that we should consider.

Reviewer: Agreed. IMO, the pipeline injection/override design can be deferred.

Author: I added a sample of what the CLI flags would look like to do an image override, I agree, having them in the config is redundant.  Fixed in the latest commit.

   * create the test Pod
   * when the test pod completes, fetch the test image Pod log output for aggregation of test results
 * display the test results 

### Custom Test Environments

Scorecard would allow you to run tests on a local Kube cluster as is the case today.  Provisioning the test cluster is external to the scope of the scorecard itself.  Note, that with this proposal, scorecard would not be responsible for installing OLM or depend on OLM being installed in order to run. 

#### Migrating Scorecard Internal Tests to Custom Test Images

With the above custom test capability, there is benefit to Red Hat in migrating the existing scorecard internal tests to be based on the custom test format. 

Today there are 2 internal scorecard tests that make use of the scorecard proxy, with these tests migrating to the custom test format, the scorecard proxy is likely not necessary which would also reduce the overall scorecard implementation complexity.

### Upgrade / Downgrade Strategy

The changes to scorecard are large and would require a migration off
of the current scorecard.

## Drawbacks

## Ansible and Helm Operator Support

### Ansible Operator Complex Tests

> For custom tests, it might be possible to write the custom test image using ansible's callback plugin framework to produce container image log output in the scorecard required format (e.g. v1alpha2).  The complexity of the test might dictate whether this is viable or not, the fallback would be for the custom test to be developed in python, Java, or golang as discussed in this proposal.

Reviewer 3: We should add an issue to track ansible and helm default tests. 

For ansible I assume something based on molecule, that can get the scorecard output?

/cc @fabianvf 

Not Blocking
